File: src/guild_droid/comlink.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from .models import GuildSnapshot, MemberActivity


class ComlinkError(RuntimeError):
    pass


def _integer(value: Any) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return 0


def fetch_guild(comlink_url: str, guild_id: str) -> dict[str, Any]:
    payload = json.dumps(
        {
            "payload": {
                "guildId": guild_id,
                "includeRecentGuildActivityInfo": True,
            },
            "enums": False,
        }
    ).encode("utf-8")
    request = Request(
        f"{comlink_url.rstrip('/')}/guild",
        data=payload,
        headers={"Content-Type": "application/json", "User-Agent": "BluesBrothersDroid/0.1"},
        method="POST",
    )
    try:
        with urlopen(request, timeout=45) as response:
            return json.load(response)
    except HTTPError as error:
        raise ComlinkError(f"Comlink returned HTTP {error.code}") from error
    except URLError as error:
        raise ComlinkError(f"Could not reach Comlink at {comlink_url}: {error.reason}") from error


def normalize_guild(response: dict[str, Any]) -> GuildSnapshot:
    guild = response.get("guild")
    if not isinstance(guild, dict):
        raise ComlinkError("Comlink response did not contain a guild")

    profile = guild.get("profile", {})
    members = guild.get("member", [])
    if not isinstance(profile, dict) or not isinstance(members, list):
        raise ComlinkError("Comlink returned an unexpected guild shape")

    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    inactive_cutoff_ms = now_ms - (24 * 60 * 60 * 1000)
    raid_tickets = 0
    inactive_24h = 0
    member_activity: list[MemberActivity] = []

    for member in members:
        last_activity_time = _integer(member.get("lastActivityTime"))
        if last_activity_time < inactive_cutoff_ms:
            inactive_24h += 1
        member_tickets = 0
        for contribution in member.get("memberContribution", []):
            if _integer(contribution.get("type")) == 2:
                member_tickets = _integer(contribution.get("currentValue"))
                raid_tickets += member_tickets
        member_activity.append(
            MemberActivity(
                player_id=str(member.get("playerId") or ""),
                name=str(member.get("playerName") or "Unknown player"),
                galactic_power=_integer(member.get("galacticPower")),
                raid_tickets=member_tickets,
                last_activity_time=last_activity_time,
                guild_join_time=_integer(member.get("guildJoinTime")),
            )
        )

    return GuildSnapshot.create(
        guild_name=str(profile.get("name") or "Unknown guild"),
        members=len(members),
        galactic_power=sum(_integer(member.get("galacticPower")) for member in members),
        character_power=sum(
            _integer(member.get("characterGalacticPower")) for member in members
        ),
        ship_power=sum(_integer(member.get("shipGalacticPower")) for member in members),
        galactic_legends=None,
        raid_tickets=raid_tickets,
        inactive_24h=inactive_24h,
        member_activity=member_activity,
    )


def fetch_snapshot(comlink_url: str, guild_id: str) -> GuildSnapshot:
    return normalize_guild(fetch_guild(comlink_url, guild_id))

File: web/lib/comlink.ts
import { createHash, createHmac } from "node:crypto";

type ComlinkMember = {
  playerId?: string;
  playerName?: string;
  galacticPower?: string | number;
  characterGalacticPower?: string | number;
  shipGalacticPower?: string | number;
  guildJoinTime?: string | number;
  lastActivityTime?: string | number;
  playerLevel?: string | number;
  memberLevel?: string | number;
  squadPower?: string | number;
  lifetimeSeasonScore?: string | number;
  leagueId?: string;
  guildXp?: string | number;
  memberContribution?: Array<{ type?: string | number; currentValue?: string | number }>;
};

export type TerritoryWarParticipant = {
  eligible?: boolean;
  joinTime?: string | number;
  memberId?: string;
  power?: string | number;
};

export type TerritoryWarGuildStatus = {
  conflictStatus?: Array<{
    defeatedSquadCount?: string | number;
    squadCapacity?: string | number;
    squadCount?: string | number;
    warSquad?: Array<Record<string, unknown>>;
    zoneStatus?: {
      commandMessage?: string;
      score?: string | number;
      zoneId?: string;
      zoneState?: string | number;
    };
  }>;
  profile?: {
    guildGalacticPower?: string | number;
    id?: string;
    name?: string;
  };
};

export type TerritoryWarStatus = Record<string, unknown> & {
  awayGuild?: TerritoryWarGuildStatus;
  currentRound?: string | number;
  currentRoundEndTime?: string | number;
  definitionId?: string;
  excludedFromWar?: boolean;
  homeGuild?: TerritoryWarGuildStatus;
  instanceId?: string;
  optedInMember?: TerritoryWarParticipant[];
  optedInMemberId?: string[];
};

export type TerritoryWarResult = Record<string, unknown> & {
  endTimeSeconds?: string | number;
  opponentGuildProfile?: TerritoryWarGuildStatus["profile"];
  opponentScore?: string | number;
  power?: string | number;
  score?: string | number;
  startTime?: string | number;
  territoryWarId?: string;
};

type ComlinkResponse = {
  guild?: {
    profile?: { name?: string };
    member?: ComlinkMember[];
    recentTerritoryWarResult?: TerritoryWarResult[];
    territoryWarStatus?: TerritoryWarStatus[];
  };
};

export type GuildRosterMember = {
  playerId: string;
  name: string;
  galacticPower: bigint;
  characterPower: bigint;
  shipPower: bigint;
  raidTickets: number;
  joinedAt: Date;
  lastActivityAt: Date | null;
  playerLevel: number;
  memberRole: "Member" | "Officer" | "Leader";
  squadPower: number;
  lifetimeSeasonScore: bigint;
  leagueId: string | null;
  guildXp: number;
  rawPayload: Record<string, unknown>;
};

export type GuildRoster = {
  guildId: string;
  name: string;
  members: GuildRosterMember[];
  territoryWarResults: TerritoryWarResult[];
  territoryWars: TerritoryWarStatus[];
  rawPayload: Record<string, unknown>;
};

function numberValue(value: string | number | undefined) {
  const parsed = Number(value ?? 0);
  return Number.isFinite(parsed) ? parsed : 0;
}

function dateValue(value: string | number | undefined) {
  const parsed = numberValue(value);
  if (!parsed) return null;
  return new Date(parsed < 1_000_000_000_000 ? parsed * 1000 : parsed);
}

function memberRole(value: string | number | undefined): GuildRosterMember["memberRole"] {
  if (value === 4 || value === "4" || value === "GUILD_LEADER") return "Leader";
  if (value === 3 || value === "3" || value === "GUILD_OFFICER") return "Officer";
  return "Member";
}

function authorizationHeaders(path: string, body: string): Record<string, string> {
  const accessKey = process.env.COMLINK_ACCESS_KEY;
  const secretKey = process.env.COMLINK_SECRET_KEY;
  if (!accessKey || !secretKey) return {};

  const timestamp = Date.now().toString();
  const bodyHash = createHash("md5").update(body).digest("hex");
  const signature = createHmac("sha256", secretKey)
    .update(timestamp)
    .update("POST")
    .update(path)
    .update(bodyHash)
    .digest("hex");

  return {
    "X-Date": timestamp,
    Authorization: `HMAC-SHA256 Credential=${accessKey},Signature=${signature}`,
  };
}

async function postComlink<T>(path: string, payload: Record<string, unknown>): Promise<T> {
  const baseUrl = process.env.COMLINK_URL;
  if (!baseUrl) throw new Error("COMLINK_URL is required");

  const body = JSON.stringify({ payload, enums: false });
  const response = await fetch(`${baseUrl.replace(/\/$/, "")}${path}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "User-Agent": "BluesBrothersDroid/0.7",
      ...authorizationHeaders(path, body),
    },
    body,
    signal: AbortSignal.timeout(90_000),
  });

  if (!response.ok) throw new Error(`Comlink returned HTTP ${response.status} for ${path}`);
  return (await response.json()) as T;
}

export async function fetchGuildRoster(): Promise<GuildRoster> {
  const guildId = process.env.SWGOH_GUILD_ID;
  if (!guildId) throw new Error("SWGOH_GUILD_ID is required");

  const payload = await postComlink<ComlinkResponse>("/guild", {
    guildId,
    includeRecentGuildActivityInfo: true,
  });
  const guild = payload.guild;
  if (!guild || !Array.isArray(guild.member)) {
    throw new Error("Comlink returned an unexpected guild response");
  }

  const members = guild.member.flatMap((member) => {
    const playerId = String(member.playerId ?? "").trim();
    if (!playerId) return [];
    const contributions = Array.isArray(member.memberContribution) ? member.memberContribution : [];
    const raidTickets = contributions
      .filter((item) => numberValue(item.type) === 2)
      .reduce((total, item) => total + numberValue(item.currentValue), 0);

    return [{
      playerId,
      name: String(member.playerName || "Unknown player"),
      galacticPower: BigInt(Math.trunc(numberValue(member.galacticPower))),
      characterPower: BigInt(Math.trunc(numberValue(member.characterGalacticPower))),
      shipPower: BigInt(Math.trunc(numberValue(member.shipGalacticPower))),
      raidTickets,
      joinedAt: dateValue(member.guildJoinTime) ?? new Date(),
      lastActivityAt: dateValue(member.lastActivityTime),
      playerLevel: Math.trunc(numberValue(member.playerLevel)),
      memberRole: memberRole(member.memberLevel),
      squadPower: Math.trunc(numberValue(member.squadPower)),
      lifetimeSeasonScore: BigInt(Math.trunc(numberValue(member.lifetimeSeasonScore))),
      leagueId: member.leagueId ? String(member.leagueId) : null,
      guildXp: Math.trunc(numberValue(member.guildXp)),
      rawPayload: member as Record<string, unknown>,
    }];
  });

  return {
    guildId,
    name: guild.profile?.name || "Blues Brothers",
    members,
    territoryWarResults: Array.isArray(guild.recentTerritoryWarResult) ? guild.recentTerritoryWarResult : [],
    territoryWars: Array.isArray(guild.territoryWarStatus) ? guild.territoryWarStatus : [],
    rawPayload: payload as Record<string, unknown>,
  };
}

type ComlinkPlayerResponse = {
  name?: string;
  playerId?: string;
  allyCode?: string | number;
};

export type ComlinkPlayer = {
  playerId: string;
  name: string;
};

type PlayerRosterUnit = {
  definitionId?: string;
  relic?: { currentTier?: string | number };
  purchasedAbilityId?: string[];
};

export type ComlinkPlayerProfile = Record<string, unknown> & {
  allyCode?: string | number;
  playerId?: string;
  level?: string | number;
  lifetimeSeasonScore?: string | number;
  selectedPlayerPortrait?: { id?: string };
  selectedPlayerTitle?: { id?: string };
  rosterUnit?: PlayerRosterUnit[];
  datacron?: unknown[];
};

const GALACTIC_LEGEND_IDS = new Set([
  "GLREY",
  "SUPREMELEADERKYLOREN",
  "GRANDMASTERLUKE",
  "SITHPALPATINE",
  "JEDIMASTERKENOBI",
  "LORDVADER",
  "JABBATHEHUTT",
  "GLLEIA",
  "GLAHSOKATANO",
  "GLHONDO",
]);

export function summarizePlayerProfile(profile: ComlinkPlayerProfile) {
  const roster = Array.isArray(profile.rosterUnit) ? profile.rosterUnit : [];
  const baseIds = roster.map((unit) => String(unit.definitionId ?? "").split(":")[0]);
  return {
    galacticLegends: baseIds.filter((id) => GALACTIC_LEGEND_IDS.has(id)).length,
    unlockedUltimates: roster.reduce(
      (count, unit) => count + (unit.purchasedAbilityId ?? []).filter((id) => id.startsWith("ultimateability_")).length,
      0,
    ),
    relicUnits: roster.filter((unit) => numberValue(unit.relic?.currentTier) > 2).length,
    rosterUnits: roster.length,
    datacrons: Array.isArray(profile.datacron) ? profile.datacron.length : 0,
  };
}

export async function fetchPlayerProfileById(playerId: string): Promise<ComlinkPlayerProfile> {
  return postComlink<ComlinkPlayerProfile>("/player", { playerId });
}

export function sanitizeAllyCode(input: string) {
  const digits = input.replace(/\D/g, "");
  return digits.length === 9 ? digits : null;
}

export async function fetchPlayerByAllyCode(allyCode: string): Promise<ComlinkPlayer | null> {
  const clean = sanitizeAllyCode(allyCode);
  if (!clean) return null;

  const payload = await postComlink<ComlinkPlayerResponse>("/player", { allyCode: clean });
  const playerId = String(payload.playerId ?? "").trim();
  if (!playerId) return null;

  return { playerId, name: String(payload.name || "Unknown player") };
}
